### PR TITLE
build: support mounting files to directories

### PR DESCRIPTION
**Pull request by jazzdan**
My one big question: is there a better way to indicate that the destination of a mount is a directory other than requiring that a `/` be appended?

Also I'm pretty sure this approach won't work in Windows.

### CODE REVIEW

File: internal/build/path_mapping.go [hunk 1]
```diff
@@ -111,15 +113,40 @@ func fileToPathMapping(file string, mounts []model.Mount) (pathMapping, *PathMap
...
+				ContainerPath: containerPath,
 			}, nil
 		}
 	}
 	return pathMapping{}, pathMappingErrf("file %s matches no mounts", file)
 }
 
+// TODO(dmiller): this won't work in Windows
+func endsWithSlash(path string) bool {
+	return strings.HasSuffix(path, "/")
```
Comment (nicks) [inline internal/build/path_mapping.go]: Use filepath.Separator?

File: internal/build/path_mapping.go [hunk 2]
```diff
@@ -111,15 +113,40 @@ func fileToPathMapping(file string, mounts []model.Mount) (pathMapping, *PathMap
...
+}
+
+func isFile(path string) (bool, error) {
```
Comment (nicks) [inline internal/build/path_mapping.go]: This isn't quite what you want. First, it's buggy (this treats symlinks as non-files). I'm also not sure it's what you want semantically. What does docker do if you COPY dir1 dir2/? I thought it created dir2/dir1

File: internal/build/tar.go [hunk 3]
```diff
@@ -132,10 +132,14 @@ func (a *ArchiveBuilder) tarPath(ctx context.Context, source, dest string) error
 			header.Name = strings.TrimPrefix(path, source)
 			// ...and live inside `dest`
 			header.Name = filepath.Join(dest, header.Name)
+		} else if strings.HasSuffix(dest, "/") { // TODO(dmiller): this won't work in Windows
```
Comment (nicks) [inline internal/build/tar.go]: Ditto here re: separator

File: internal/build/path_mapping.go (hunk 2 above)
Comment (jazzdan) [reply internal/build/path_mapping.go]: Good point about symlinks! Will fix.

I didn't change the handling of copying directories to directories. This should only take effect if the source is a file but the destination is a directory.

> What does docker do if you COPY dir1 dir2/? I thought it created dir2/dir1
Looks like it doesn't nest directories in that case.

```dockerfile
FROM iron/go:dev

RUN mkdir /remotedir2
COPY dir1 /remotedir2/
```

```
> tree
.
├── Dockerfile
├── dir1
│   └── hi
```

```
> docker run -it 303a3d0f2157 /bin/sh
/go # cd /tmp/
/tmp # ls
/tmp # cd /
/ # ls
bin         etc         home        media       proc        root        sbin        sys         usr
dev         go          lib         mnt         remotedir2  run         srv         tmp         var
/ # cd remotedir2/
/remotedir2 # ls
hi
```

Comment (nicks) [reply internal/build/path_mapping.go]: i would probably change this to be !isDir instead of isFile, which is probably closer to what we want?

**Pull request merged**